runtime/error.c

```diff
@@ -148,15 +161,71 @@ void _bal_error_backtrace_print(ErrorPtr ep) {
     fflush(stderr);
 }
 
+static void processPCs(GC PC* pcs, uint32_t nPCs, uint32_t start, int64_t lineNumber, BacktraceError *error) {
+    // Handle the starting pc seperately
+    BacktraceStartLine backtraceStartLine = { { error->code, error->message }, lineNumber };
+    processInitialPC(pcs[start], lineNumber, &backtraceStartLine);
+    backtraceStartLine.lineNumber = 0;
+    if (backtraceStartLine.error.code == NO_DEBUG_INFO) {
+        for (uint32_t i = start + 1; i < nPCs; i++) {
+            // We do not need to pass error_callback because, already we have an error.
+            backtrace_syminfo(state, pcs[i], printMissingSymbols, NULL, &backtraceStartLine);
+        }
+    }
+    else {
+        for (uint32_t i = start + 1; i < nPCs; i++) {
+            backtrace_pcinfo(state, pcs[i], printBacktraceLine, storeBacktraceError, &backtraceStartLine);
+        }
+    }
+    error->code = backtraceStartLine.error.code;
+    error->message = backtraceStartLine.error.message;
+}
+
+static void processInitialPC(PC pc, int64_t lineNumber, BacktraceStartLine *backtraceStartLine) {
+    backtrace_pcinfo(state, pc, printBacktraceLine, storeBacktraceError, backtraceStartLine);
+    if (backtraceStartLine->error.code == NO_DEBUG_INFO) {
+        // We do not need to pass error_callback because, already we have an error
+        backtrace_syminfo(state, pc, printMissingSymbols, NULL, backtraceStartLine);
+    }
+}
+
 // Implementation of backtrace_full_callback
 static int printBacktraceLine(void *data, PC pc, const char *filename, int lineno, const char *function) {
-    if (function != NULL && filename != NULL) {
-        fprintf(stderr, "%s %s:%d\n", function, filename, lineno);
-        fflush(stderr);
+    BacktraceStartLine *backtraceStartLine = data;
+    if (backtraceStartLine->lineNumber != 0) {
+        lineno = backtraceStartLine->lineNumber;
     }
+    printLine(filename, lineno, function);
     return SUCCESS;
 }
 
+// Implementation of backtrace_syminfo_callback
+static void printMissingSymbols(void *data, uintptr_t pc, const char *symname, uintptr_t symval, uintptr_t symsize) {
+    BacktraceStartLine *backtraceStartLine = data;
+    printLine(NULL, backtraceStartLine->lineNumber, symname);
+}
+
+static void printLine(const char *filename, int64_t lineNumber, const char *function) {
+    // API docs say any of filename, lineno and function may be 0 meaning unavailable
+    const char *sep = "    ";
+    if (function != NULL) {
+        fprintf(stderr, "%s%s", sep, function);
```

Comment:
I think it's better to use compile time string concatenation. 
```C
define ERR_SEPR "     "
...

fprintf(stderr, "%s" ERR_SEPR, function);
```

Reply:
Oh I see you are assigning to `sep` below, hmm still looks wired to me, why concat static strings at runtime?

Reply:
I wrote this: it keeps the cases more uniform. There's no performance issue here.